### PR TITLE
Add an option to remember last opened Extensions in Dynamo

### PR DESCRIPTION
**Pull request by zeusongit**
### Purpose

https://jira.autodesk.com/browse/DYN-5570
This PR adds an option to remember last opened Extensions, a new toggle button is added under Features > Extensions section of preference panel to enable/disable this feature.
When enabled, the user will see view extensions which were open while closing Dynamo in the last session.
When disabled, dynamo will not open extensions from the prior session or save their state.
The extensions are serialized in the settings XML file.

Unit tests added -
- When option is enabled, checks if extensions that were open while closing Dynamo during the last session do reopen.
- When option is enabled, checks if extensions that were closed while closing Dynamo during the last session do not reopen. 
- When option is disabled, checks if extensions that were open while closing Dynamo during the last session do not reopen.

Snapshot:
DynamoSandbox_UeDuVULa4i

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Release Notes

- Add an option to remember last opened Extensions in Dynamo

### Reviewers

@DynamoDS/dynamo

### CODE REVIEW

File: src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs [hunk 1]
```diff
@@ -1,28 +1,26 @@
...
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.IO;
+using System.Linq;
+using System.Security.Permissions;
+using System.Threading;
+using System.Windows;
+using System.Windows.Controls;
+using System.Windows.Threading;
+using DynamoProperties = Dynamo.Properties;
```
Comment (QilongTang) [inline src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs:23]: Please follow https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards#usings

File: src/DynamoCore/Configuration/PreferenceSettings.cs [hunk 2]
```diff
@@ -655,7 +672,7 @@ public string SelectedPackagePathForInstall
         /// <summary>
         /// Settings that apply to view extensions.
         /// </summary>
-        public List<ViewExtensionSettings> ViewExtensionSettings { get; set; }
+        public List<ViewExtensionSettings> ViewExtensionSettingsList { get; set; }
```
Comment (QilongTang) [inline src/DynamoCore/Configuration/PreferenceSettings.cs:675]: Would this cause old settings to not deserialize correctly right?

File: src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs [hunk 3]
```diff
@@ -771,7 +771,7 @@ public string SelectedPythonEngine
         }
         
         /// <summary>
-        /// Controls the IsChecked property in the "Hide IronPython alerts" toogle button
+        /// Controls the IsChecked property in the "Hide IronPython alerts" toggle button
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs:774]: Nice

File: src/WorkspaceDependencyViewExtension/WorkspaceDependencyViewExtension.cs [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
 using Dynamo.Core;
 using Dynamo.Extensions;
 using Dynamo.Graph.Workspaces;
 using Dynamo.Logging;
 using Dynamo.PackageManager;
 using Dynamo.WorkspaceDependency.Properties;
 using Dynamo.Wpf.Extensions;
+using System;
+using System.Linq;
+using System.Windows.Controls;
```
Comment (QilongTang) [inline src/WorkspaceDependencyViewExtension/WorkspaceDependencyViewExtension.cs:10]: same comments here

File: src/DynamoCore/Configuration/PreferenceSettings.cs (hunk 2 above)
Comment (mjkkirschner) [reply src/DynamoCore/Configuration/PreferenceSettings.cs:675]: this is an API break.

File: src/DynamoCore/Configuration/ViewExtensionSettings.cs [hunk 5]
```diff
@@ -21,6 +21,10 @@ public class ViewExtensionSettings
         /// Window settings for the extension control when displayed in FloatingWindow mode.
         /// </summary>
         public WindowSettings WindowSettings { get; set; }
+        /// <summary>
+        /// Specifies if the extension was Open in the last session before closing Dynamo, if the property to remember view extension status was enabled.<br/> <c>Default: False</c>
+        /// </summary>
+        public bool IsOpen { get; set; } = false;
```
Comment (mjkkirschner) [inline src/DynamoCore/Configuration/ViewExtensionSettings.cs:27]: probably redundant ... bool should default to false

File: src/DynamoCore/Configuration/PreferenceSettings.cs (hunk 2 above)
Comment (zeusongit) [reply src/DynamoCore/Configuration/PreferenceSettings.cs:675]: Ok will revert, but this make our xml look like this:

```
< ViewExtensionSettings >
< ViewExtensionSettings></ViewExtensionSettings>
< ViewExtensionSettings></ViewExtensionSettings>
</ViewExtensionSettings>
```

Parent and child nodes have the same name.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 6]
```diff
@@ -1599,14 +1627,43 @@ private void WindowClosing(object sender, CancelEventArgs e)
...
+                SaveExtensionOpenState(window);
+            }
+            //for any new extensions that are opened for the first time
+            foreach (var tab in ExtensionTabItems)
+            {
+                SaveExtensionOpenState(tab);
+            }
+            //update open state of all existing view extension in setting, if option to remember extensions is enabled in preferences
+            if (dynamoViewModel.PreferenceSettings.EnablePersistExtensions) {
+                var settings = dynamoViewModel.PreferenceSettings.ViewExtensionSettingsList;
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:1648]: maybe just always do this - and just let the preference control loading?
If I am not mistaken you are checking the setting here and at load time?

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 7]
```diff
@@ -1599,14 +1627,43 @@ private void WindowClosing(object sender, CancelEventArgs e)
...
+                        setting.IsOpen = false;
+                    }
+                }
+            }
+        }
+        //This method is to ensure that the extensions states are correctly saved for newly added extensions.
+        private void SaveExtensionOpenState(object o)
+        {
+            if (!dynamoViewModel.PreferenceSettings.EnablePersistExtensions || o == null) return;
+            var extId = o is TabItem ? ((TabItem)o).Uid : ((ExtensionWindow)o).Uid;
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:1662]: this should probably be a type switch or other pattern match with a default case that fails gracefully with a message.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 8]
```diff
@@ -1599,14 +1627,43 @@ private void WindowClosing(object sender, CancelEventArgs e)
...
+            var setting = dynamoViewModel.Model.PreferenceSettings.ViewExtensionSettingsList.Find(ext => ext.UniqueId == extId);
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:1663]: I think some `?` null checks are in order.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 9]
```diff
@@ -63,6 +63,7 @@ namespace Dynamo.Controls
     public partial class DynamoView : Window, IDisposable
     {
         public const string BackgroundPreviewName = "BackgroundPreview";
+        private const int SideBarCollapseThreshold = 20;
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:66]: this is actually a nice thing to add to the preferences!

File: src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs [hunk 10]
```diff
@@ -179,8 +177,7 @@ public override void Loaded(ViewLoadedParams viewLoadedParams)
             this.ViewModel.DynamoView = viewLoadedParams.DynamoWindow;
         }
 
-
-        public void Shutdown()
+        public override void Shutdown()
```
Comment (mjkkirschner) [inline src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs:180]: was the compiler complaining about method hiding?

File: src/DynamoCore/Configuration/PreferenceSettings.cs (hunk 2 above)
Comment (mjkkirschner) [reply src/DynamoCore/Configuration/PreferenceSettings.cs:675]: if you want to make this change you should figure out how to handle the old deserialization and keep the old property working/obsolete it - you don't know who is using it.

File: src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs (hunk 1 above)
Comment (zeusongit) [reply src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs:23]: I did not know this setting, thanks.

**Comment by mjkkirschner** [comment]
looks pretty solid @zeusongit - a few commnts.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs (hunk 7 above)
Comment (zeusongit) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:1662]: message in logger?

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs (hunk 6 above)
Comment (zeusongit) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:1648]: Yes, at load time to display extensions and here to update them.
Can you elaborate on `always do this`?

Comment (mjkkirschner) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:1648]: always update them, no matter what state the preference is in.

File: src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs (hunk 10 above)
Comment (zeusongit) [reply src/DocumentationBrowserViewExtension/DocumentationBrowserViewExtension.cs:179]: yes, there was a warning.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs (hunk 6 above)
Comment (zeusongit) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs]: So basically remove the `EnablePersistExtensions` property check?

Comment (mjkkirschner) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs]: yeah, just an idea to simplify it - is it needed?

Comment (zeusongit) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs]: not needed, will update, won't hurt to default to false.

**Comment by mjkkirschner** [comment]
some test failures on latest commits.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 11]
```diff
@@ -1599,14 +1627,57 @@ private void WindowClosing(object sender, CancelEventArgs e)
...
+                    break;
+                case ExtensionWindow w:
+                    extId = w.Uid;
+                    break;
+                default:
+                    Log("Incorrect extension type, could not save extension state.");
+                    break;
+            }
+
+            if (extId == "") return;
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:1675]: Please use string.empty

**Review by QilongTang — APPROVED** [review]
LGTM with couple comments

File: src/DynamoCore/Configuration/PreferenceSettings.cs [hunk 12]
```diff
@@ -575,6 +576,22 @@ public bool EnableNotificationCenter
             }
         }
 
+        /// <summary>
+        /// This defines if user wants to the Extensions settings to persist across sessions.
```
Comment (QilongTang) [inline src/DynamoCore/Configuration/PreferenceSettings.cs:580]: `This defines if user wants the Extensions settings to persist across sessions.`

**Comment by QilongTang** [comment]
Thanks @zeusongit Since I see the previous Dynamo SelfServe check already passed, will merge once basic ones passing

**Pull request merged**